splice: Script for complex splices

ddustin opened the pull request:
PR reviewers: It might be good to start here to learn the splice script syntax: https://github.com/ddustin/lightning/blob/ddustin/splice_script/doc/lightning-splice.7.md

New compiler for splice scripts that parses splice scripts, validates them, converts them to json and back again.

A test is included that puts some advanced and complex scripts into the parser and validates the results that come out.

Changelog-Added: Splice script parser — takes a custom splice query language to bundle multiple complex splices into a single task in a simple way.

vincenzopalazzo reviewed (commented):
Mh What do you mean by "Scripts"? and why now we need a "compiler" for splicing? 😕

I feel that I am missing somethings?

ddustin commented:
> Mh What do you mean by "Scripts"? and why now we need a "compiler" for splicing? 😕
> 
> 
> 
> I feel that I am missing somethings?

Proper documentation is coming but here's a quick summary:

The script isn't required to be used -- there will also be a json option and, of course, the lower level api is available.

When doing splices across multiple channels it quickly becomes unwieldy for users. Using the script is a nice shorthand to make it easier -- and also allows you to do "dry runs" and confirm it's doing what you'd like.

Say you wanted to take 50% of funds and splice them into three other channels, in splice script it's quite simple:

```
84ff50 -> 50%;
* -> cecf35;
* -> 4fe310;
* -> c14f60;
```

While working on it, it made sense to add a bunch of other useful things like being able to specify amounts as whole numbers, percentages, or * (split amount among all *s); channel queries; onchain bitcoin addresses; and the onchain wallet.

By building this more powerful part first it becomes trivial to implement the simpler splice commands. For instance:

Splice out 10 million sats
```
84ff50 -> 10M;
```

Splice in 10 million sats
```
wallet -> 10M;
* -> 84ff50;
```

Cross splice 10 million sats
```
84ff50 -> 10M;
* -> cecf35;
```

Pay 10M sats onchain using channel funds
 ```
84ff50 -> 10M+fee;
10M -> 1JfbZRwdDHKZmuiZgYArJZhcuuzuw2HuMu;
```

ddustin commented:
You can see a description of what it does here:
https://github.com/ElementsProject/lightning/blob/ab42c231624664e33ee4437f2b3028b519aa8f53/doc/lightning-splice.7.md

chrisguida commented:
Anyone know what's up with this backslash in the docs for `user_provided_sats`? It's both in the markdown preview and in the manpage, but it appears to be named the right thing in [the schema](https://github.com/ElementsProject/lightning/pull/6980/files#diff-f792af080490603d3090a65c1a64115b9eb9003a3ec6254fbca5caff3ca4261aR27), so maybe something is just off in the docs rendering?

![image](https://github.com/ElementsProject/lightning/assets/7445670/0578e5d7-38ad-4bd0-9d1a-440b00f1ca16)

```
SYNOPSIS
       (WARNING: experimental-splicing only)

       splice [script] [dryrun] [json] [psbt] [user_provided_\sats]
```

chrisguida commented:
I can't seem to get past this:

```
$ l1-cli splice '039751b -> 10000sat' true
{
   "code": -32602,
   "message": "Splice script compile failed",
   "compiler_error": [
      "Missing channel or address equivilent. Each line must contain one",
      "039751b -> 10000sat",
      "           ^",
      "Compiler phase: make_segments"
   ]
}
```

What am I doing wrong?

chrisguida commented:
Oh hmm I guess it wanted the channel_id not the peer_id.

Now it's just hanging...

chrisguida commented:
If I do `dryrun` then it doesn't hang

```
$ l1-cli splice "c0a77695 -> 10M;" true
{
   "dryrun": [
      "c0a77695f94848c330f62bb1b7112ba2bf153e5f221c0155becc43299415f93d withdraw 10000000sat",
      "put 100% of rest into wallet"
   ]
}
```

chrisguida commented:
Hmm now it says 

```
$ l1-cli splice "c0a77695 -> 10000sat;"
{
   "code": 355,
   "message": "Currently waiting on previous splice command to finish."
}
```

I guess the old splice is still hanging somewhere?

ddustin commented:
Thanks for reviewing!

> ```
> SYNOPSIS
>        (WARNING: experimental-splicing only)
> 
>        splice [script] [dryrun] [json] [psbt] [user_provided_\sats]
> ```

Ah it was a typo, nice spotting.

> Oh hmm I guess it wanted the `channel_id` not the `peer_id`.
> 
> Now it's just hanging...

Yeah it needs a channel id or a channel query. If you want to use a node id you can specify "the first channel we have with node id" by saying `039751b:0` or `039751b:?`

Might be worth making a clearer error message for this case 🤔

> ```
> $ l1-cli splice "c0a77695 -> 10000sat;"
> {
>    "code": 355,
>    "message": "Currently waiting on previous splice command to finish."
> }
> ```
> 
> I guess the old splice is still hanging somewhere?

Yeah that happened because a previous splice got stuck prior or is waiting for user input. Restarting the node will clear it.

If you remember which command made the splice get stuck that would be helpful 🙏

chrisguida commented:
This was the one that hung

```
$ l1-cli splice "c0a77695 -> 10M;"
^C
```

Maybe because I didn't have that much in the channel? I just kinda slapped that in there without thinking much, guessing that's a splice out?

chrisguida commented:
Ok I tried it with a smaller amount and it seems to have worked!

```
$ l1-cli splice "c0a77695 -> 10000sat;"
{
   "psbt": "cHNidP8BAgQCAAAAAQQBAQEFAQIBBgEDAfsEAgAAAAABAP21AQIAAAAAAQJ1c2jX5HBTkzsEVasZnFd4tRbXb9QY+Ws1nnfQYjTVzwAAAAAA/f///4H8KHX0JJL0Na5sxWcu78ag+sW6MGl3Sajj9fBwgBQYAAAAAAD9////A/BZ9AUAAAAAIlEg5bV970yfonIJ0MpBPnJvpxi29LLnS3GLwK12WqUpMHkanuYFAAAAACJRIKZ9GbEKjvEgVWwI0hhGSOftkX7Q4lFrGyf+2PzsFHOd4MgQAAAAAAAiACDLZac0GecjZnnNTqJg5oadKounzGzU5eAM+uXGGQlgrQJHMEQCIFlmZJRHMig5YihgBwnuDvXGHn9PD9LzEjPDTWnlIxvbAiARUxLInnqYASliMbIQsvDpaTqkDgVJ2C1bBDzizu81kwEhA7iCzk9M8Npoyx2DEo38CkGsoGIJ46kA3gq+pnzgRS6sAkcwRAIgS10MidyLHnu0TV0meRGVWHsf7gwXvpEg+pktGbp0+/8CIBWKvQizYlnpcExtrJcpXdlpubRZFmBYwFC37X+HyVNfASEDLpZHCRF9B+n4HD7Je9QItJwNd5ub8kxh6mZN7ohZEEz0AAAAAQEr4MgQAAAAAAAiACDLZac0GecjZnnNTqJg5oadKounzGzU5eAM+uXGGQlgrSICAo6BpOai25V7gdMYxODX1ykKBhUPqHdSJ/YAAHr/V8DRRzBEAiBAQBWz4ulhD4k2TYdE/kDl7cs9+0emd95mx90Wob/hcwIgEePQiTkU0BD4s9mvKdc1JLZaUnKqnQzOI9z604m5GwABAQMEAQAAAAEOIPC7IXX1IS6zUoQvcjR6AVGJPDdrkI6UGT/oRdXQ4tURAQ8EAgAAAAEQBAAAAAAM/AlsaWdodG5pbmcBCLTD7bTesGowAAEDCDYmAAAAAAAAAQQiUSCLlzxf8KSQryCUGZhXUCSooXsoo1KCaIGjAFmj7JIspwz8CWxpZ2h0bmluZwEIQgenL51TwwQAAQMI0KEQAAAAAAABBCIAIMtlpzQZ5yNmec1OomDmhp0qi6fMbNTl4Az65cYZCWCtDPwJbGlnaHRuaW5nAQiZvH8MEdONxAA=",
   "txid": "4a282a73dc411412f27d4de02ab4fad87e398207164c49fb3a907599f3adc8f8"
}
```

chrisguida commented:
Interesting, this is still here even though I mined 6 blocks and the channel is back to normal

```
         "inflight": [
            {
               "funding_txid": "4a282a73dc411412f27d4de02ab4fad87e398207164c49fb3a907599f3adc8f8",
               "funding_outnum": 1,
               "feerate": "253perkw",
               "total_funding_msat": 1090000000,
               "our_funding_msat": 1000000000,
               "splice_amount": -10000,
               "scratch_txid": "5a8909f2114723d754c9415a0a09b925487e293a5c972545751cc9a7da05cb52"
            }
         ],
```

How long does this output stick around?

ddustin commented:
> This was the one that hung
> 
> ```
> $ l1-cli splice "c0a77695 -> 10M;"
> ^C
> ```
> 
> Maybe because I didn't have that much in the channel? I just kinda slapped that in there without thinking much, guessing that's a splice out?
> 
> Edit: ah yeah, docs say "Splice out 10 million satoshis to the onchain wallet" xD
> 
> Might be worth it to return some kind of error message instead of hanging though?

Yeah it should definitely be throwing an error in this case.



> Interesting, this is still here even though I mined 6 blocks and the channel is back to normal
> 
> ```
>          "inflight": [
>             {
>                "funding_txid": "4a282a73dc411412f27d4de02ab4fad87e398207164c49fb3a907599f3adc8f8",
>                "funding_outnum": 1,
>                "feerate": "253perkw",
>                "total_funding_msat": 1090000000,
>                "our_funding_msat": 1000000000,
>                "splice_amount": -10000,
>                "scratch_txid": "5a8909f2114723d754c9415a0a09b925487e293a5c972545751cc9a7da05cb52"
>             }
>          ],
> ```
> 
> How long does this output stick around?

It shouldn't be -- nice catch.

chrisguida commented:
Yeah I guess I instinctively tried a peer id because I was used to [this behavior](https://docs.corelightning.org/reference/lightning-close):

>If the given id is a peer ID (66 hex digits as a string), then it
applies to the active channel of the direct peer corresponding to the
given peer ID. If the given id is a channel ID (64 hex digits as a
string, or the short channel ID blockheight:txindex:outindex form),
then it applies to that channel.

niftynei commented on common/splice_script.c:
```diff
@@ -0,0 +1,2696 @@
+
+		if (fee) {
+			itr->pays_fee = true;
+			itr->feerate_per_kw = fee->amount_sat.satoshis; /* Raw: We use satoshis as our base number unit */
```
little weird?

ddustin replied:
Duplicate of https://github.com/ElementsProject/lightning/pull/6980#discussion_r1569444636

niftynei commented on common/splice_script.c:
```diff
@@ -0,0 +1,2696 @@
+							 "calculate_amounts");
+			}
+			if (input[i]->right->type == TOK_PERCENT)
+				right_used_ppm += input[i]->right->ppm;
```
nit: overflow check?

ddustin replied:
`right_used_ppm` is only used to determine if there is any non-zero ppm amount so I believe this is fine.

ddustin replied:
I'll switch it to a bool

niftynei reviewed (commented):
review checkpoint!

niftynei commented on common/splice_script.c:
```diff
@@ -0,0 +1,2696 @@
+						   struct token **tokensB,
+						   size_t b_size)
+{
+	for (size_t i = 0; i < channels_count; i++) {
```
sort before going thru it?? global sort on stash?

niftynei replied:
todo: make a new method on the channel fetcher (wallet.c) that always returns sorted channels by oldest to newest.

ddustin replied:
Duplicate of https://github.com/ElementsProject/lightning/pull/6980#discussion_r1621356185

niftynei commented on common/splice_script.c:
```diff
@@ -0,0 +1,2696 @@
+{
+	for (size_t i = 0; i < channels_count; i++)
+		if (node_id_eq(&node_id, &channels[i].node_id))
+			if (!channel_index--)
```
can we use == 0? please

niftynei commented on common/splice_script.c:
```diff
@@ -0,0 +1,2696 @@
+					 "resolve_channel_ids");
+		case TOK_LEASERATE:
+		case TOK_ARROW:
+		case TOK_STR:
```
should be gone too?

niftynei commented on common/splice_script.c:
```diff
@@ -0,0 +1,2696 @@
+					 "segments");
+		case TOK_ARROW:
+		case TOK_PIPE:
+		case TOK_STR:
```
should be in above section?

niftynei commented on common/splice_script.c:
```diff
@@ -0,0 +1,2696 @@
+		case TOK_FEERATE:
+		case TOK_NODEID:
+		case TOK_BTCADDR:
+		case TOK_CHANQUERY:
```
should be in section above?

niftynei reviewed (commented):
checkpoint!

niftynei commented on plugins/spender/splice.c:
```diff
@@ -0,0 +1,1402 @@
+		if (splice_cmd->actions[i]->out_ppm)
+			return do_fail(cmd, splice_cmd, JSONRPC2_INVALID_PARAMS,
+					    "Should be no out_ppm on final");
+		if (splice_cmd->actions[i]->pays_fee && pays_fee++)
```
use BOOL!!

niftynei commented on plugins/spender/splice.c:
```diff
@@ -0,0 +1,1402 @@
+			tok.start = 0;
+			tok.end = strlen(action->bitcoin_address);
+			tok.size = tok.end;
+			switch(json_to_address_scriptpubkey(cmd,
```
this is so hacky. we can have nice things! just make a bitcoin address -> scriptpubkey method?

ddustin replied:
I began the work on making an address -> scriptpubkey function but it's incomplete.

Moving it to a new task: https://github.com/ElementsProject/lightning/issues/5911

niftynei commented on plugins/spender/splice.c:
```diff
@@ -0,0 +1,1402 @@
+						    scriptpubkey,
+						    action->in_sat);
+
+			/* DTODO: should be action->in_sat but we haven't calculated those yet... */
```
what does this mean??

ddustin replied:
Updated to more meaningful `DTODO: support dynamic address payouts (percent)`

niftynei commented on plugins/spender/splice.c:
```diff
@@ -0,0 +1,1402 @@
+			else {
+				return do_fail(cmd, splice_cmd,
+					       JSONRPC2_INVALID_PARAMS,
+					       "Putting change back into same"
```
this error message doesn't seem to match the crime?

niftynei commented on plugins/spender/splice.c:
```diff
@@ -0,0 +1,1402 @@
+		if (action->in_ppm) {
+			/* ppm percentage calculation:
+			 * action->in_sat = out_sats * in_ppm / 1000000 */
+			if (!amount_sat_mul(&action->in_sat, out_sats, action->in_ppm))
```
can we assert that `in_sat` isn't set?

niftynei commented on plugins/spender/splice.c:
```diff
@@ -0,0 +1,1402 @@
+					    onchain_fee))
+				return do_fail(cmd, splice_cmd,
+						    JSONRPC2_INVALID_PARAMS,
+						    "Unable to sub fee from"
```
nit: "unable to *add* fee to.."

niftynei commented on plugins/spender/splice.c:
```diff
@@ -0,0 +1,1402 @@
+		}
+
+		/* If this item pays the fee, subtract it from either their
+		 * in_sats or out_sats. */
```
*or add to `out_sats`

niftynei commented on plugins/spender/splice.c:
```diff
@@ -0,0 +1,1402 @@
+	size_t extra_inputs = 0;
+	size_t extra_outputs = 0;
+
+	/* BOLT-0d8b701614b09c6ee4172b04da2203e73deec7e2 #2:
```
nit: just use BOLT #2 ??

niftynei commented on plugins/spender/splice.c:
```diff
@@ -0,0 +1,1402 @@
+}
+
+static struct command_result *splice_update_get_result(struct command *cmd,
+			      const char *buf,
```
start here!

niftynei commented on plugins/spender/splice.c:
```diff
@@ -0,0 +1,1402 @@
+	*index = UINT32_MAX;
+	for (size_t i = 0; i < tal_count(splice_cmd->actions); i++) {
+		if (splice_cmd->states[i]->state == SPLICE_CMD_UPDATE_DONE) {
+			/* There can only be one node that requires our sigs */
```
you need sigs to be a Directed Acyclic Graph (DAG); which you can get by requiring the more-sats side to send second?

ddustin replied:
Yup that should be handled by the lower layer (channeld)

niftynei commented on plugins/spender/splice.c:
```diff
@@ -0,0 +1,1402 @@
+	for (size_t i = 0; i < tal_count(splice_cmd->actions); i++) {
+		if (splice_cmd->states[i]->state == SPLICE_CMD_UPDATE_DONE) {
+			/* There can only be one node that requires our sigs */
+			assert(!action);
```
i dont think an assert is the right way to handle this.

niftynei commented on plugins/spender/splice.c:
```diff
@@ -0,0 +1,1402 @@
+				    signpsbt_get_result, splice_error,
+				    splice_cmd);
+
+	/* TODO: add the inputs indices we should sign */
```
THIS IS VERY IMPORTANT, right?

ddustin replied:
Adding to phase two splicing issue: https://github.com/ElementsProject/lightning/issues/5911

ddustin replied:
Think this over, it is too important to move to another phase. I added input tracking to 195a73c6d0a78847b4010b56fc5fa11b30949461

niftynei commented on plugins/spender/splice.c:
```diff
@@ -0,0 +1,1402 @@
+	/* TODO: add the inputs indices we should sign */
+	json_add_psbt(req->js, "psbt", splice_cmd->psbt);
+
+	splice_cmd->wallet_inputs_to_signed = 0;
```
delete this??

niftynei commented on channeld/channeld.c:
```diff
@@ -3074,14 +3075,20 @@ static struct amount_sat check_balances(struct peer *peer,
 	max_initiator_fee = amount_tx_fee(peer->feerate_max,
 					  calc_weight(TX_INITIATOR, psbt));
+	penalty_fee = amount_tx_fee(peer->feerate_penalty,
+				    calc_weight(TX_ACCEPTER, psbt));
```
TX_ACCEPTER -> should be what your role is, no?

niftynei commented on contrib/pyln-client/pyln/client/lightning.py:
```diff
@@ -1148,6 +1148,20 @@ def openchannel_abort(self, channel_id):
         }
         return self.call("openchannel_abort", payload)
 
+    def splice(self, script, dryrun=None, psbt=None, user_provided_sats=None, force_feerate=None, json=None, debug_log=None, wetrun=None):
```
this is cool!!

ddustin replied:
lets goooo!!!

niftynei commented on tests/test_splice.py:
```diff
@@ -0,0 +1,60 @@
+@unittest.skipIf(TEST_NETWORK != 'regtest', 'elementsd doesnt yet support PSBT features we need')
+def test_script_splice_out(node_factory, bitcoind):
+    l1, l2 = node_factory.line_graph(2, fundamount=1000000, wait_for_announce=True, opts={'experimental-splicing': None})
+    l1.rpc.splice("*:? -> 100000; 100%-fee -> wallet", force_feerate=True, debug_log=True)
```
a comment saying what this is testing would be very nice.

niftynei commented on tests/test_splice.py:
```diff
@@ -0,0 +1,60 @@
+    assert p2['to_us_msat'] == 0
+    assert p2['total_msat'] == 900000000
+    assert 'inflight' not in p1
+    assert 'inflight' not in p2
```
check that the wallet balance has updated?  `listfunds` or `bkpr-listbalances` ?

niftynei commented on tests/test_splice.py:
```diff
@@ -0,0 +1,60 @@
+    assert p2['to_us_msat'] == 0
+    assert p2['total_msat'] == 1100000000
+    assert 'inflight' not in p1
+    assert 'inflight' not in p2
```
could add the fees as a check? maybe check the bkpr- commands that might be able to give you a total cost of fees for a channel?

ddustin replied:
Can we push this to the second pass?

ddustin replied:
Adding to the splicing issue: https://github.com/ElementsProject/lightning/issues/5911

ddustin commented on common/test/run-splice_script.c:
```diff
@@ -0,0 +1,310 @@
+
+	assert(i == tal_count(final));
+
+	printf("DRY RUN:\n%s",
```
I believe outputting is happening in other tests and the output during normal test operation gets discard to /dev/null anyway

ddustin commented on lightningd/channel_control.c:
```diff
@@ -2271,6 +2364,103 @@ static struct command_result *json_splice_signed(struct command *cmd,
+		subd_send_msg(channels[i]->owner,
+			      take(towire_channeld_abort(NULL)));
+
+	return command_success(cmd, json_stream_success(cmd));
```
Hm what use case would this be for? I suppose I figured abort is ending the process so not a lot of value in getting something back.

ddustin commented on wallet/reservation.c:
```diff
@@ -751,11 +751,193 @@ static const struct json_command addpsbtoutput_command = {
+			/* Uneconomic to add this utxo, skip it */
+			if (!all && amount_sat_greater_eq(fee, utxo->amount))
+				continue;
+			if (!all && utxo_is_csv_locked(utxo, current_height))
```
Ya I think we shouldn't include them

ddustin commented on common/splice_script.c:
```diff
@@ -0,0 +1,2696 @@
+struct splice_script_error *parse_splice_script(const tal_t *ctx,
+						const char *script,
+						struct splice_script_chan *channels,
+						size_t channels_count,
```
I switched it over to dynamic allocations & tals.

ddustin commented on common/splice_script.c:
```diff
@@ -0,0 +1,2696 @@
+				if (fee->type != TOK_FEE)
+					return new_error(ctx, INTERNAL_ERROR,
+							 fee,
+							 "validate_and_clean");
```
great idea 👍

ddustin commented on plugins/spender/splice.c:
```diff
@@ -0,0 +1,1402 @@
+			splice_cmd->states[i]->state = SPLICE_CMD_NONE;
+		}
+	}
+
```
These dont appear to apply to plugins

ddustin commented on lightningd/channel_control.c:
```diff
@@ -2170,51 +2178,95 @@ static struct command_result *json_splice_update(struct command *cmd,
+	subd_send_msg(channel->owner, take(msg));
+	if (success)
+		*success = true;
+	return command_still_pending(cmd);
```
```
struct command_result *command_still_pending(struct command *cmd)
{
	notleak_with_children(cmd);
	cmd->pending = true;

	/* If we've started writing, wake reader. */
	if (cmd->json_stream)
		json_stream_flush(cmd->json_stream);

	return &pending;
}
```
Looks pretty safe to call multiple times to me. `notleak_with_children` is just for dev diagnostics.

ddustin commented on lightningd/channel_control.c:
```diff
@@ -2271,6 +2364,103 @@ static struct command_result *json_splice_signed(struct command *cmd,
+			 NULL))
+		return command_param_failed();
+
+	channels = tal_arr(cmd, struct channel*, 0);
```
The two places we load the channels (`stfu` and `abort`) have slightly different needs. The `stfu` one needs to check if the channel has an active splice command and error out, while `abort` doesn't need or want this check.

I think this reduces the value of cutting it into its own function since the channels will need to be iterated anyway.

ddustin commented on wallet/reservation.c:
```diff
@@ -751,11 +751,193 @@ static const struct json_command addpsbtoutput_command = {
+			utxo_weight = utxo_spend_weight(utxo, 0);
+			fee = amount_tx_fee(*min_feerate, utxo_weight);
+
+			/* Uneconomic to add this utxo, skip it */
```
It's mimicking the `utxopsbt` command. I believe this is the currently correct (tm) way to do it 🤔

ddustin commented on doc/lightning-splice.7.md:
```diff
@@ -0,0 +1,371 @@
+--------
+**(WARNING: experimental-splicing only)**
+
+**splice** [*script*] [*dryrun*] [*psbt*] [*user\_provided\_sats*] [*force\_feerate*] [*json*] [*debug\_log*] [*wetrun*]
```
> * can you consolidate `script` and `json` into a single input that's required? parse one way, and then the other. fail if both fail.

Done 👍
> * does the command check only obsolete the `dryrun` command?

command check only doesn't appear to work in plugins
> * can you make `wetrun` be `dev-wetrun` and only allow it as an option if in developer mode?

Done 👍
> * todo for nifty: point dustin to library about how to return in-progress information from a command?

json_notify doesn't appear to work in plugins

ddustin commented on doc/lightning-splice.7.md:
```diff
@@ -0,0 +1,371 @@
+will be completed and published on chain.
+
+However if you specified your own inputs using `psbt`, then the command will
+return a psbt that you must sign and then pass to `splice_signed`.
```
> suggestion: can we pull this out of this PR and add it back in the next? so no `psbt` option and `user_provided_sats` this command.

Done 👍

ddustin commented on doc/lightning-splice.7.md:
```diff
@@ -0,0 +1,371 @@
+
+Another way to specify channel ids is a channel query. These take the form of a
+node id, a colon, and a number or symbol (ie. "034645dc:0") which results in
+your first channel with the peer whose node id starts with 034645dc. Change the
```
It relies on `listpeerchannels`. Perhaps the right approach is make that RPC calls ordering more strict?

ddustin commented on doc/lightning-splice.7.md:
````diff
@@ -0,0 +1,371 @@
+"wallet" operator can be placed just like channel identifiers. We can take funds
+out of the wallet like so:
+```shell
+wallet -> 10000sat
````
Can we mark this as a todo for the second pass?

ddustin replied:
Adding to the splicing issue: https://github.com/ElementsProject/lightning/issues/5911

ddustin commented on common/splice_script.c:
```diff
@@ -0,0 +1,2696 @@
+static struct token *new_token(const tal_t *ctx, enum token_type token_type,
+			       size_t script_index)
+{
+	struct token *token = talz(ctx, struct token);
```
I prefer doing it manually because each of these `new_token_*` functions would only be used once. I also like when the code shows `struct->param = value` so `param` is visible as opposed to hidden by the function call -- which is extra useful when doing tokens where `param`s have a tendency to get overloaded.

ddustin replied:
Got rid of the talz usage and explicitly set values to 0

ddustin commented on common/splice_script.c:
```diff
@@ -0,0 +1,2696 @@
+
+				script_index++;
+
+				tokens[n] = new_token(tokens, TOK_STR,
```
Discussing it here: https://github.com/ElementsProject/lightning/pull/6980#discussion_r1621358569

ddustin replied:
Going to mark this one as resolved since it's a duplicate

ddustin commented on common/splice_script.c:
```diff
@@ -0,0 +1,2696 @@
+		return false;
+	if (len > PUBKEY_CMPR_LEN)
+		return false;
+	if (!hex_decode(token->str, len * 2,
```
Did you want to match half-bytes here?

I'm not sure if an odd hex digit is the upper half of the byte or the lower half. I think it may not be defined

Most of the internet seems to just reject odd number hex strings. Google addresses odd number of hex digits by prepending an extra 0 on the front.
![IMAGE 2024-05-30 16:24:01](https://github.com/ElementsProject/lightning/assets/331415/bc20f33e-f5e0-4289-ad3f-bb8eb774d4e7)
![IMAGE 2024-05-30 16:24:09](https://github.com/ElementsProject/lightning/assets/331415/ed86fa3b-6ad4-4561-92c7-47657490ea50)

Which would definetely *not* work well for channel identifiers

If I have a channel
`0399069f1693fd89a453f0caf03ee36b6f6c8abaa7ef778d3e2bcc7c2b44120100`
And I querty it with
`03990`
I definetely do not mean to search for 
`003990`

So we'd have to invent a new spec for hex encoding which would probably go against consensus of how they are used.

I prefer dropping the rightmost digit for odd numbered hex strings as then `03990` would match our example.

ddustin commented on lightningd/channel_control.c:
```diff
@@ -2170,51 +2178,95 @@ static struct command_result *json_splice_update(struct command *cmd,
+		success = false;
+		result = single_splice_signed(cmd, channels[i], psbt,
+					      *sign_first, &success);
+		/* If one channel fails, we stop there and return the error */
```
Might be worth considering this mode as an option in the future 🤔

ddustin replied:
Will move this to the phase two issue https://github.com/ElementsProject/lightning/issues/5911

ddustin commented on common/splice_script.c:
```diff
@@ -0,0 +1,2696 @@
+				chan_id = chan_for_node_index(ctx, channels,
+							      channels_count,
+							      *input[i]->left->node_id,
+							      input[i]->right->amount_sat.satoshis); /* Raw: We use satoshis as index values in splice script */
```
Duplicate of https://github.com/ElementsProject/lightning/pull/6980#discussion_r1569444636

ddustin commented on bitcoin/psbt.c:
```diff
@@ -311,6 +311,36 @@ bool psbt_input_have_signature(const struct wally_psbt *psbt,
+	if (ok) {
+		item = &psbt->inputs[in].signatures.items[index_plus_one - 1];
+		*sig = tal(ctx, struct bitcoin_signature);
+		if (!signature_from_der(item->value, item->value_len, *sig)) {
```
`struct bitcoin_signature` only supports ecdsa currently. I'll rename the method to be more clear it only supports ecdsa

ddustin commented on common/splice_script.c:
```diff
@@ -0,0 +1,2696 @@
+		return NULL;
+}
+
+static bool is_bitcoin_address(const char *address)
```
Adding to splicing issue https://github.com/ElementsProject/lightning/issues/5911

niftynei commented on lightningd/channel_control.c:
```diff
@@ -2382,6 +2389,10 @@ static struct command_result *json_stfu_channels(struct command *cmd,
+	if (command_check_only(cmd))
+		return command_check_done(cmd);
+
+	stfu_result = tal_arr(cmd->ld, struct stfu_result*, 0);
```
add comment about lifetime of `stfu_result`?

niftynei reviewed (commented):
there's a few changes to be made and fixups need to be squashed + rebase needed but giving it a thumbs up for now!

ddustin commented:
Fix the doc newline issue

ddustin commented on common/addr.c:
```diff
@@ -36,3 +36,79 @@ char *encode_scriptpubkey_to_addr(const tal_t *ctx,
+			    scriptpubkey_p2sh_hash(ctx, &destination.addr);
+			return true;
+		} else {
+			return false;
```
Doing in second phase, added to issue: https://github.com/ElementsProject/lightning/issues/5911

niftynei reviewed (approved):
ACK ac1cabc831b2f38883ebbdb220cab3f0044e8519

rustyrussell commented:
OK, can we make this experimental, or a dev- command?  I'm not convinced this is the way we want to go,  but it's *great* for exercising all the nasty cases, and may well be useful for larger nodes.

Needs rebase because it was now "added v24.11" not "v24.08", sorry :(

ddustin commented:
> OK, can we make this experimental, or a dev- command? I'm not convinced this is the way we want to go, but it's _great_ for exercising all the nasty cases, and may well be useful for larger nodes.
> 
> Needs rebase because it was now "added v24.11" not "v24.08", sorry :(

I changed the command to "dev-splice," updated the added tag to 24.11, and fixed a handful of issues that came up from rebasing.

Can we push this one through once it passes CI @rustyrussell?

rustyrussell commented:
Ack!  Will apply once CI is happy!

rustyrussell commented:
Accidentally told GH to update to latest master and it did a merge.  Gah!  Rebased instead.
